cmd/zgenhostid: replace with simple c implementation

gyakovlev opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
In #10887 it was
discovered that dracut scripts and zgenhostid
always generate little-endian binary /etc/hostid.

### Description
<!--- Describe your changes in detail -->

This commit provides simple endianess-aware binary.

New features include:
 `-f` flag to force overwrite.
 `-o` flag to write to different file (for dracut)
 acceping both `0x01234567` and `01234567` formats as input
Manpage updated accordingly.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

I haven't written a line of `C` in many years, this needs a good review.

Tested by generating `/tmp/hostid` on different systems, including musl libc, which does not provide `sethostid()`
Also tested by passing arbitrary values, seems to be working ok.

```bash-session
#little-endian
./zgenhostid -o /tmp/hostid -f 01234567
hexdump -ve '1/1 "%02X"' /tmp/hostid
67452301
```

```bash-session
#big-endian
./zgenhostid -o /tmp/hostid -f 01234567
hexdump -ve '1/1 "%02X"' /tmp/hostid
01234567

#another test
zgenhostid -o /tmp/hostid "$(hostid)" && diff -s /etc/hostid /tmp/hostid
Files /etc/hostid and /tmp/hostid are identical
```

also tested dracut

it properly reacts to `spl_hostid` on command line and properly copies `hostid` into initrd.

```shell-session
...
[    4.895210] dracut: ZFS: Using hostid from command line: 0x01234567    
...                                                                                                                                         
# zdb -C zroot | grep hostid
        hostid: 19088743
echo 'ibase=10;obase=16;19088743' |bc
1234567
```
### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

gyakovlev commented:
apparently `0xDEADBEEF` is not the best string to test with =)
```
./zgenhostid -f -o /tmp/hostid 12345678 && od -x /tmp/hostid
0000000 3412 7856
0000004
```
not swapping properly on little-endian. will look later.
force-pushed new version wil a bit of options.

behlendorf commented on cmd/zgenhostid/zgenhostid.c:
```diff
@@ -0,0 +1,146 @@
+	int opt;			// getopt iterator
+	int forcew = 0;			// force  file write
+	uint32_t hostid = 0;		// final hostid value
+	unsigned long int input_i = 0;	// int to hold converted input
```
`unsigned long int` -> `unsigned long` is sufficient here.   Both `hostid` and `input_i` could also be declared down farther where they're first used to limit their scope.

gyakovlev replied:
done  renamed some variables for clarity, moved hostid down.
I don't think I can move input_i down, as it used in 2 for loops, one that handles user input, and second time if it needs generation.

behlendorf commented on cmd/zgenhostid/zgenhostid.c:
```diff
@@ -0,0 +1,146 @@
+		}
+	}
+
+	if (argv[optind]) {
```
It might be clearer to assign  `in_s` here to `argv[optind]` and then perform the check.  I don't believe there's any need to make a copy.  For example the following, and then use `strlen()` below.

```c
char *in_s = argv[optind];
if (in_s != NULL) {
}
```

gyakovlev replied:
good call. done.

behlendorf commented on cmd/zgenhostid/zgenhostid.c:
```diff
@@ -0,0 +1,146 @@
+int
+main(int argc, char **argv)
+{
+	int opt;			// getopt iterator
```
nit: `/* */` style comments are preferred to `//` for consistency with the rest of the source.

behlendorf commented on cmd/zgenhostid/zgenhostid.c:
```diff
@@ -0,0 +1,146 @@
+		// we need exactly 8 or 10 (if input is 0x prefixed) characters
+		const char *hex = "0123456789abcdefxABCDEFX";
+		int numhex = 0;
+		numhex = strspn(in_s, hex);
```
nit: we can declare and assign on the same line.

```
int numhex = strspn(in_s, hex);
```

behlendorf commented on cmd/zgenhostid/zgenhostid.c:
```diff
@@ -0,0 +1,146 @@
+#include <time.h>
+#include <unistd.h>
+
+static void usage(void);
```
nit: I'd suggest moving the `usage()` function before `main()` to avoid the need for a forward reference.

behlendorf commented on cmd/zgenhostid/zgenhostid.c:
```diff
@@ -0,0 +1,146 @@
+			exit(EXIT_FAILURE);
+		}
+
+		if (input_i < 0x1 || input_i > 0xFFFFFFFF) {
```
You could use `UINT32_MAX` here.

behlendorf reviewed (commented):
For not writing C in a while this looks pretty good.  I've added a few minor suggestions.  Thanks for tackling this!

gyakovlev commented:
addressed feedback, also added filepath lengh check. it would have been caught by fopen/fwrite, but I now handle if user passes `-o ""`
also did some minor variable renaming for clarity, while morning coffee kicked in.

ofaaland commented on cmd/zgenhostid/zgenhostid.c:
```diff
@@ -38,44 +38,71 @@
-		int numhex = 0;
-		numhex = strspn(in_s, hex);
+		int numhex = strspn(in_s, hex);
 		if (numhex != 8 && numhex != 10) {
```
I believe that if the input string starts with 10 hex digits, but has no 0x prefix, strtoul will parse 8 digits of it, potentially leading to confusion.  For example, with string "1111222233".  Since you require they give you exactly 8 digits of hex anyway, maybe just keep track of the number of prefix digits and subtract that from the length, e.g. ```if ((numhex-numprefix) != 8)```.

gyakovlev replied:
numhex may be == 10, but int range check will not succeed at least on systems where ULONG_MAX is 8 bytes.

but it can misfire on 32bit systems, because `strtoul` will return `ULONG_MAX` which will likely be equal to `UINT32_MAX`
but the *endptr check was supposed to prevent that, but looks like it's not working as it should, because condition defined wrong.

on ppc32
```
./a.out -o /tmp/hostid -f ffffffffff
0000000 ffff ffff
0000004
```
changing `if (errno != 0)` to be only condition properly catches truncation, so logic `if (errno != 0 && *endptr != '\0')` is not working.

I think it should be sufficient for catching overflows, because `errno` will be non-zero if truncated/overflown, so I'll remove endptr check.

thanks for catching!

I can skip first 2 bytes if string starts with `0x` and check exactly for 8 characters, this will simplify things.

ofaaland commented on cmd/zgenhostid/zgenhostid.c:
```diff
@@ -88,20 +115,23 @@ main(int argc, char **argv)
 	struct stat fstat;
-	if (forcew == 0 && stat(path, &fstat) == 0 &&
+	if (force_fwrite == 0 && stat(path, &fstat) == 0 &&
 	    S_ISREG(fstat.st_mode) && fstat.st_size == 4) {
```
I can imagine sysadmins not noticing the size of /etc/hostid files and being confused that sometimes zgenhostid will replace them, and other times will not.

gyakovlev replied:
so what's the preferred way? just check for presence and `S_ISREG`, removing size check?

ofaaland reviewed (commented):
Thanks for the improvement!  Sorry I didn't think of the endianness issue in the first place.  I noted a couple of corner cases I believe may lead to confusion.   I may have overlooked something, so feel free to push back.

gyakovlev commented:
Thank for your valuable feedback. thanks to @ofaaland  we found an issue with integer check on 32bit systems. fixed and simplified code by using provided suggestions.

ofaaland reviewed (approved):
Looks good, thanks again for the patch.
